### ISSUE TITLE
Create an individual p2 repository per subproject

### ISSUE THREAD

**Issue by spoenemann** [eclipse/xtext#1085]
Currently the Xtext subprojects (xtext-core, xtext-eclipse, etc.) have very different ways to publish their results: xtext-eclipse and xtext-xtend  use p2 repositories, all others use Maven repositories. As a consequence, we need an umbrella build that aggregates the build results from all subprojects and assembles a p2 repository in order to consume any part of Xtext using a target platform.

We would like to change this structure such that each subproject is self-contained both from the Maven artifacts perspective and from the p2 artifacts perspective. This means that the subprojects with a Gradle build need to trigger Tycho for assembling a local p2 repository that can be consumed in a similar way as the already existing local Maven repositories. This is already [done in LSP4J](https://github.com/eclipse/lsp4j/blob/master/gradle/p2-deployment.gradle), for example. We will still need an umbrella build to create the cross-subproject features such as Xtext SDK. Regarding the Xtext download space on Eclipse, we could either deploy the individual p2 repositories and let them reference each other, or **let the umbrella build create a single p2 repository with all features and deploy that one**.

A prerequisite for this further restructuring is that we create at least one new Eclipse feature in each subproject such that each feature contains only plug-ins from its source subproject and references features from upstream subprojects. The old features, e.g. Xtext SDK, would be changed so they no longer *contain* features and plug-ins, but *reference* features.

To summarize the main consequences:

**++** By referencing the local p2 repositories of the Jenkins jobs, you will be able to refresh a target platform in your Xtext-based project immediately after a build has finished on one of the Jenkins jobs. No triggering of additional umbrella builds will be necessary.

**&ndash; &ndash;** The restructuring of the Eclipse features might make it more complicated to update to Xtext 2.11 for existing Eclipse users. In some cases they might have to uninstall the old features and install the new ones instead of directly updating.

If you have worries regarding this feature and p2 restructuring, please comment here.

**Comment by dhuebner** [comment]
> they no longer contain features and plug-ins, but reference features

We need then reference the features by match="perfect" see http://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Ffeature_manifest.html

**Comment by svenefftinge** [comment]
Sounds good. I would prefer if the "features projects" are not contained in the normal module structure but are hidden in some releng folder, like we have it for LSP4J.

**Comment by meysholdt** [comment]
sounds good to me too. some thoughts:
* I think we can create backwards-compatible features by ensuring the features deploy the same bundles. This means: if a local feature includes a foreign (from another repo) feature, change *include* to *depends*. If a local feature includes a foreign bundle, changes this to depend on a foreign feature which then includes the foreign bundle.
* big yay to eliminating artifact-redeployment! Except for the umbrella build. There I think it is of value to have  self-contained deliverable. 
* +1 for hiding features inside a *releng* folder, just like here https://github.com/eclipse/xtext-eclipse/tree/master/releng
* I'm not sure we need to invoke the Tycho build from gradle. Invoking it from the `Jenkinsfile` may avoid *wtf*-moments among the gradle users.
* do we really need a feature for each repo that includes all local bundles? My concern is that features should be designed for users, but here our design incentives are driven by our internal technical structure. We can, however, introduce any kinds of features as long as we *do not* redeploy them in the umbrella build.

**Comment by holgerschill** [comment]
+1

**Comment by LorenzoBettini** [comment]
By "deploy the individual p2 repositories and let them reference each other" do you mean a composite update site? Because recently you can add a reference to another repository in the category.xml but IIRC Tycho still does not support that; and in any case, such information would be available only in recent versions of Eclipse.

I think it is wrong that a fetaure "depends" on other features of the same provider, instead of "including"/"containing" them (at least, that's my understanding of "depending" and "containing"). By containing a feature you also impose a "perfect" match, while if you depend on them you need to impose "perfect" match manually, as @dhuebner said. Not to mention that if you depend on something, instead of including, that something will not be published in the p2 repository.

Couldn't the umbrella projects (like, e.g., Xtext SDK) simply refer to all the other features in its target platform, include the features/plugins as it did before, and publish the whole p2 repository (which would be the "official" one)?

You can still have single subprojects publish their bits in their p2 repos, that can be consumed for development purposes. You just need to be careful and avoid running into something similar to https://github.com/eclipse/xtext-eclipse/issues/101: if Xtend IDE SDK includes Xbase lib feature, then it must take Xbase lib from Xbase p2 repo (through the target platform) and not rebuild that, otherwise you end up with two different published Xbase lib versions into two different p2 repos.

**Comment by kittaakos** [comment]
> ++ By referencing the local p2 repositories of the Jenkins jobs, you will be able to refresh a target platform in your Xtext-based project immediately after a build has finished on one of the Jenkins jobs. No triggering of additional umbrella builds will be necessary.

The sub-repository based build could also be very useful when you need to have a build with a modified version of Xtext; for instance when you are working on a branch, and you need a TP, hence a build.

+1

**Comment by spoenemann** [comment]
We could also consider to leave the features as they are and build them all in the umbrella build. Then all other subprojects could create p2 repositories that contain plain plug-ins. This approach would avoid potential problems from breaking features and also avoid designing features along our technical structure as mentioned by @meysholdt, but still allow to fetch the required stuff directly from a subproject Jenkins build job.

**Comment by cdietrich** [comment]
so we would use a "fake" catch all feature? or can tcho build non feature based p2 repos?

**Comment by spoenemann** [comment]
Yes, you can simply include bundles in a p2 repo.

**Comment by spoenemann** [comment]
See the six PRs listed above. This change does not modify the features, but makes sure that
 * the Tycho builds in xtext-eclipse, xtext-xtend, and xtext-umbrella consume only p2 repositories and no Maven dependencies,
 * xtext-umbrella is the only one who publishes content that does not come from the same repository.

This means that the main repositories (lib, core, extras, eclipse, xtend) are self-contained regarding their published artifacts. In order to accomplish this, I moved
 * `org.eclipse.xtext.examples`, `org.eclipse.xtext.redist`, `org.eclipse.xtext.runtime`, `org.eclipse.xtext.ui`, `org.eclipse.xtext.xbase`, and `org.eclipse.xtext.xtext.ui` from xtext-eclipse to xtext-umbrella
 * `org.eclipse.xtext.gwt` and `org.eclipse.xtext.xbase.lib` from xtext-eclipse to xtext-lib
 * `org.eclipse.xtend.sdk` from xtext-xtend to xtext-umbrella
 * `org.eclipse.xtend.gwt` from xtext-xtend to xtext-lib

We should discuss the restructuring of features as a second step. I would even propose to do that after the 2.11 release.

**Comment by spoenemann** [comment]
The restructuring of p2 repositories is done. I issued eclipse/xtext-umbrella#12 for the restructuring of features.
